src/Psl/IO/CloseSeekReadWriteHandle.php

```diff
@@ -0,0 +1,9 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\IO;
+
+interface CloseSeekReadWriteHandle extends CloseReadWriteHandle, CloseSeekReadHandle, CloseSeekWriteHandle, SeekReadWriteHandle
```

Comment:
Can you tell me what extra value these combined interfaces provide?
Is it something you will actually typehint on. I'dd assume if I used this in the code, I'dd typehint for `ResourceHandle` most of the time?

Reply:
most of the time you want to type hint for the feature you want to be enabled in the handle, STDOUT and others cannot be closed, if you type hint against `ResourceHandle`, then you will have access to ->close(), which doesn't actually work in that case.

Reply:
That is true, but the `open` function always returns a `ResourceHandle`. So even if you typehint it as a `WriteHandle`, you still have access to the close() function.

Reply:
`ResourceHandle` is internal, and you shouldn't type hint against it, so is `open`.

e.g:

```
my_function(ReadHandle $handle): void {
  // do something with it
}

$stdout = IO\output_handle();
my_function($stdout); // error: expected `ReadHandle` got `WriteHandle`

$stdin = IO\input_handle();
my_function($stdin); // ok
``` 

i know at runtime this will not trigger an error ( at least for the type ), but i'm not sure how to handle that.

IMO, we can handle a decorator for each interface, where the decorator takes `ResourceHandle` and only implement functions for said interface, so that at runtime, you will also get a type error.

e.g:

```
/**
 * @internal
 */
final class WriteHandleDecorator implements WriteHandle {
  public function __construct(private ResourceHandle $handle) {}

  // implement write() here
}
```

so we can have `IO\input_handle` and friends return decorator instances that will ensure no misuse happens.

WDYT?

Reply:
ow, missed the fact that you marked `ResourceHandle` as internal. However that doesn't really matter, since it all is publicly accessible.

The big question is : do you want to put in the extra effort for runtime safety or do you trust the users to have static analyzers in place that will tell them they are doing it wrong. 

For me personally, it doesnt matter since psalm will tell me I'm wrong. But I don't the majority of projects use that yet.

I do like the decorators approach: that would also give an actual use-case for the interfaces you added.

Reply:
i think we should put the effort, since even psalm will agree with this:

```
function f(ReadHandle $handle): void {
  if ($handle instanceof SeekHandle) {
    $handle->seek(0);
  }

  $data = $handle->read();
}
```

if you pass `IO\input_handle` to `f`, it shouldn't be able to seek through it, but currently, since the return value is `ResourceHandle`, it can, and both runtime and psalm will agree with it ( even tho that seek will through an exception ).

Reply:
hm, figured i can do:

```php
function input_handle(): ReadHandle
{
    if (PHP_SAPI === "cli") {
        $handle = Internal\open('php://stdin', 'rb');
    } else {
        $handle = Internal\open('php://input', 'rb');
    }

    return new class($handle) implements ReadHandle {
        private Internal\ResourceHandle $handle;

        public function __construct(
            Internal\ResourceHandle $handle
        ) {
            $this->handle = $handle;
        }

        /**
         * An immediate, unordered read.
         *
         * @param int|null $max_bytes the maximum number of bytes to read.
         *
         * up to `$max_bytes` may be allocated in a buffer; large values may lead
         * to unnecessarily hitting the request memory limit.
         *
         * @returns string the read data on success, or an empty string if the end of file is reached.
         *
         * @throws Exception\BlockingException if there is no more data available to read.
         * @throws InvariantViolationException If $max_bytes is 0.
         */
        public function read(?int $max_bytes = null): string
        {
            return $this->handle->read($max_bytes);
        }
    };
}
```

is there any downside to this that didn't cross my mind?

Reply:
The biggest issue imo is reusability: you now always have to repeat that method and it's docblock if you want to create a read-only handle.

Reply:
fixed this, added 2 internal decorators.